### ISSUE TITLE
ChecksumList should have ability to self validate (replace verifyCsigChecksumFile/verifyCsigChecksumList)

### ISSUE THREAD

**Issue by heddn** [drupal/php-signify#16]
```
$files = $verifier->verifyCsigMessage($message): \string;
$checksum_list = new ChecksumList($files, TRUE);
$failures_iterator = $checksum_list->getFailedFiles(): \Iterator;
$all_files = $checksum_list->getAllFiles(): \Iterator;
```

**Comment by mbaynton** [comment]
The code sample calls one `verify*` method, and it currently returns a string. I expected `$files` to be a string, with higher-level value objects coming from `new ChecksumList`. For clarification, does this issue request `verifyCsigMessage` return a value object?

**Comment by pwolanin** [comment]
I'm not sure what the problem is here - you want to add a method to verify the files list in one call?

**Comment by mbaynton** [comment]
Heddn can't use verifyCsigChecksum* with the preflight check because it throws on first patched file. This separates the signature verification step from verifying file hashes. That way signature verifications continue throwing exceptions on failures, while relaxing that policy for on-disk file hash verification.

**Comment by heddn** [comment]
Because the existing interface provided by `verifyCsigChecksumList` and requires the following work around:

```php
    $failed_files = [];
    $files = $this->verifier->verifyCsigMessage($message);
    $checksum_list = new ChecksumList($files, TRUE);
    foreach ($checksum_list as $checksum) {
      $hash_file_path = $this->drupalRoot . DIRECTORY_SEPARATOR . $checksum->filename;
      $algorithm = strtolower($checksum->algorithm);
      $hash = @hash_file($algorithm, $hash_file_path);
      if ($hash !== $checksum->hex_hash) {
        $failed_files = $checksum->filename);
      }
    }
    // Do something w/ $failed_files
```

**Comment by heddn** [comment]
Thinking about implementation, I'd want to use [FilterIterator](https://www.php.net/manual/en/class.filteriterator.php). This would work well with accepting a stream as input into the constructor, instead of a string.

**Comment by pwolanin** [comment]
I don't think this is a work-around.  You are doing exactly what we expected you to do.

The library should not support the preflight use case directly

**Comment by heddn** [comment]
I need it all the time. If `Node.php` and `Term.php` have a critical security fix and the aren't modified/patched/hacked, then we should be able to update those files. Even if `action.module` was patched for an unrelated feature.

**Comment by pwolanin** [comment]
The use case for the library is to verify the download of an archive file with an intermediate signature, or to extract a message with an intermediate signature.

The preflight use case is out of scope for the library to support directly.

**Comment by mbaynton** [comment]
> The preflight use case is out of scope for the library to support directly.

Sort of. @davidstrauss and I cite the fact that since Signify supports the full task of verifying a file hash list, it makes sense for that processing to occur in this library. Given that, well, it is in actuality `drupal/php-signify` and its primary user is going to be Drupal, I personally think it's okay to be receptive to Drupal's needs. Otherwise the alternative is to duplicate similar code for verifying the file hash list outside of this library in Drupal.

**Comment by heddn** [comment]
This was fixed in #17.
